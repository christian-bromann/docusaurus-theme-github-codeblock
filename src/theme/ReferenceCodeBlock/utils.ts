import { initialFetchResultState } from './constants'
import type { GitHubReference, DispatchMessage } from '../types'

/**
 * parses GitHub reference
 * @param {string} ref url to github file
 */
export function parseReference(ref: string): GitHubReference {
    const fullUrl = ref.slice(ref.indexOf('https'))
    const [url, loc] = fullUrl.split('#')

    /**
     * webpack causes failures when it tries to render this page
     */
    const [org, repo, blob, branch, ...pathSeg] = new global.URL(url).pathname.split('/').slice(1)
    const [fromLine, toLine] = loc
        ? loc.split('-').map((lineNr) => parseInt(lineNr.slice(1), 10) - 1)
        : [0, Infinity]

    return {
        url: `https://raw.githubusercontent.com/${org}/${repo}/${branch}/${pathSeg.join('/')}`,
        fromLine,
        toLine,
        title: pathSeg.join('/'),
        org,
        repo
    }
}

export async function fetchCode({ url, fromLine, toLine }: GitHubReference, fetchResultStateDispatcher: React.Dispatch<DispatchMessage>) {
    let res: Response

    try {
        res = await fetch(url)
    } catch (err) {
        return fetchResultStateDispatcher({ type: 'error', value: err as Error })
    }

    if (res.status !== 200) {
        const error = await res.text()
        return fetchResultStateDispatcher({ type: 'error', value: error })
    }

    const body = (await res.text()).split('\n').slice(fromLine, (toLine || fromLine) + 1)
    const preceedingSpace = body.reduce((prev: number, line: string) => {
        if (line.length === 0) {
            return prev
        }

        const spaces = line.match(/^\s+/)
        if (spaces) {
            return Math.min(prev, spaces[0].length)
        }

        return 0
    }, Infinity)

    if (body.length === 0) {
        return fetchResultStateDispatcher({
            type: 'error',
            value: `Error: No code found at ${url} from line ${fromLine} to line ${toLine}`
        })
    }

    return fetchResultStateDispatcher({
        type: 'loaded',
        value: body.map((line) => line.slice(preceedingSpace)).join('\n')
    })
}

export function codeReducer(prevState: any, { type, value }: DispatchMessage) {
    switch (type) {
        case 'reset': {
            return initialFetchResultState;
        }
        case 'loading': {
            return { ...prevState, loading: true };
        }
        case 'loaded': {
            return { ...prevState, code: value, loading: false };
        }
        case 'error': {
            return { ...prevState, code: value, loading: false };
        }
        default:
            return prevState;
    }
}
