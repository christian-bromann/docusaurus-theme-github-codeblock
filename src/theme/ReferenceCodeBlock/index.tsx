import React, { useReducer } from 'react'
import CodeBlock from '@theme-init/CodeBlock'
import { useCodeblockThemeConfig } from 'docusaurus-theme-github-codeblock/client'

import { RunmeLink } from './RunmeLink.js'
import { parseReference, fetchCode, codeReducer } from './utils.js'
import { initialFetchResultState, buttonStyles } from './constants.js'
import type { ReferenceCodeBlockProps } from '../types'

const buttonBarStyles: React.CSSProperties = {
    fontSize: '.9em',
    fontWeight: 600,
    color: '#0E75DD',
    textAlign: 'right',
    paddingBottom: '13px',
    textDecoration: 'underline'
}

function ReferenceCode(props: ReferenceCodeBlockProps) {
    const [fetchResultState, fetchResultStateDispatcher] = useReducer(
        codeReducer,
        initialFetchResultState,
    )

    const codeSnippetDetails = parseReference(props.children)
    if (fetchResultState.loading !== false) {
        fetchCode(codeSnippetDetails, fetchResultStateDispatcher)
    }

    const titleMatch = props.metastring?.match(/title="(?<title>[^"]*)"/);
    const customProps = {
        ...props,
        metastring: titleMatch?.groups?.title
            ? `${props.metastring} title="${titleMatch?.groups?.title}"`
            : `${props.metastring} title="${codeSnippetDetails.title}"`,
        children: initialFetchResultState.code,
    };

    const codeblockConfig = useCodeblockThemeConfig()
    const showButtons = codeblockConfig.showGithubLink || codeblockConfig.showRunmeLink
    return (
        <div className='docusaurus-theme-github-codeblock'>
            <CodeBlock {...customProps}>{fetchResultState.code}</CodeBlock>
            {showButtons && (
                <div style={buttonBarStyles}>
                    <RunmeLink
                        reference={props.children}
                        metastring={props.metastring}
                    />
                    {codeblockConfig.showGithubLink && (
                        <a
                            href={props.children}
                            className='githubLink'
                            style={buttonStyles}
                            target="_blank"
                        >
                            {codeblockConfig.githubLinkLabel}
                        </a>
                    )}
                </div>
            )}
        </div>
    );
}

export default ReferenceCode;
