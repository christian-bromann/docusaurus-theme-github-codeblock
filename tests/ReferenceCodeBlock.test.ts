/**
 * @jest-environment jsdom
 */
import { getRunmeLink } from '../src/theme/ReferenceCodeBlock/RunmeLink'
import { parseReference, codeReducer } from '../src/theme/ReferenceCodeBlock/utils'

const exampleLink = 'https://github.com/christian-bromann/docusaurus-theme-github-codeblock/blob/main/src/theme/ReferenceCodeBlock/index.tsx'

test('should parse GitHub reference properly', () => {
    expect(parseReference(exampleLink)).toMatchSnapshot()
    expect(parseReference(`${exampleLink}#L105-L108`)).toMatchSnapshot()
})

test('codeReducer', () => {
    const prevState = { foo: 'bar' }
    expect(codeReducer(prevState, { type: 'reset', value: '' })).toMatchSnapshot()
    expect(codeReducer(prevState, { type: 'loading', value: '' })).toMatchSnapshot()
    expect(codeReducer(prevState, { type: 'loaded', value: 'foobar' })).toMatchSnapshot()
    expect(codeReducer(prevState, { type: 'error', value: 'ups' })).toMatchSnapshot()
    // @ts-expect-error
    expect(codeReducer(prevState, { type: 'unknown', value: '' })).toEqual(prevState)
})

test('getRunmeLink', () => {
    expect(getRunmeLink(exampleLink, '')).toMatchSnapshot()
    expect(getRunmeLink(exampleLink, 'reference useHTTPS')).toMatchSnapshot()
    expect(getRunmeLink(exampleLink, 'reference useHTTPS=true')).toMatchSnapshot()
    expect(getRunmeLink(exampleLink, 'reference useHTTPS=false')).toMatchSnapshot()
    expect(getRunmeLink(exampleLink, 'runmeFileToOpen="foobar" title="barfoo"')).toMatchSnapshot()
    expect(getRunmeLink(exampleLink, 'runmeRepository="foobar" title="barfoo" runmeFileToOpen="foofoo"')).toMatchSnapshot()
})
