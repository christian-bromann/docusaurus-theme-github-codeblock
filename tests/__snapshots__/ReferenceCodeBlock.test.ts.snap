// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codeReducer 1`] = `
{
  "code": "loading...",
  "error": null,
  "loading": null,
}
`;

exports[`codeReducer 2`] = `
{
  "foo": "bar",
  "loading": true,
}
`;

exports[`codeReducer 3`] = `
{
  "code": "foobar",
  "foo": "bar",
  "loading": false,
}
`;

exports[`codeReducer 4`] = `
{
  "code": "ups",
  "foo": "bar",
  "loading": false,
}
`;

exports[`getRunmeLink 1`] = `"command=setup&repository=git%40github.com%3Achristian-bromann%2Fdocusaurus-theme-github-codeblock.git&fileToOpen=src%2Ftheme%2FReferenceCodeBlock%2FREADME.md"`;

exports[`getRunmeLink 2`] = `"command=setup&repository=https%3A%2F%2Fgithub.com%2Fchristian-bromann%2Fdocusaurus-theme-github-codeblock.git&fileToOpen=src%2Ftheme%2FReferenceCodeBlock%2FREADME.md"`;

exports[`getRunmeLink 3`] = `"command=setup&repository=https%3A%2F%2Fgithub.com%2Fchristian-bromann%2Fdocusaurus-theme-github-codeblock.git&fileToOpen=src%2Ftheme%2FReferenceCodeBlock%2FREADME.md"`;

exports[`getRunmeLink 4`] = `"command=setup&repository=git%40github.com%3Achristian-bromann%2Fdocusaurus-theme-github-codeblock.git&fileToOpen=src%2Ftheme%2FReferenceCodeBlock%2FREADME.md"`;

exports[`getRunmeLink 5`] = `"command=setup&fileToOpen=foobar"`;

exports[`getRunmeLink 6`] = `"command=setup&fileToOpen=foofoo&repository=foobar"`;

exports[`should parse GitHub reference properly 1`] = `
{
  "fromLine": 0,
  "org": "christian-bromann",
  "repo": "docusaurus-theme-github-codeblock",
  "title": "src/theme/ReferenceCodeBlock/index.tsx",
  "toLine": Infinity,
  "url": "https://raw.githubusercontent.com/christian-bromann/docusaurus-theme-github-codeblock/main/src/theme/ReferenceCodeBlock/index.tsx",
}
`;

exports[`should parse GitHub reference properly 2`] = `
{
  "fromLine": 104,
  "org": "christian-bromann",
  "repo": "docusaurus-theme-github-codeblock",
  "title": "src/theme/ReferenceCodeBlock/index.tsx",
  "toLine": 107,
  "url": "https://raw.githubusercontent.com/christian-bromann/docusaurus-theme-github-codeblock/main/src/theme/ReferenceCodeBlock/index.tsx",
}
`;
